components/brave_new_tab_ui/apiEventsToStore.ts

```diff
@@ -48,6 +48,14 @@ export function wireApiEventsToStore () {
       rewardsInitData()
     }
     getActions().setInitialData(initialData)
+    if (!initialData.preferences.widgetVisibilityMigrated) {
+      // Other widget's auth state can be fetched from local storage.
+      // But, need to fetch ftx auth state to migrate.
+      // Start migration when ftx auth state is ready.
+      chrome.ftx.getAccountBalances((balances, authInvalid) => {
+       getActions().setWidgetVisibilityMigrationData(!authInvalid)
+      })
```

Comment:
I'm wondering if we can keep the frontend simpler, by doing all the logic here without another redux action (since 'setWidgetVisibilityMigrationData' is not really an event which needs to be handled). And also not adding to app state or the `new_tab_storage.ts` which should just be limited to the stored persistent state. We have all the data we need in this function via `initialData` and we can modify it before it is sent to the store. Something like this small change to where your functions are called and the data they use?:

```ts
export function wireApiEventsToStore () {
  // Get initial data and dispatch to store
  getInitialData()
  .then(async (initialData) => {
    if (initialData.preferences.showRewards) {
      rewardsInitData()
    }
    if (!initialData.preferences.widgetVisibilityMigrated) {
      // Alternatively, this can be done inside `getInitialData` function
      await migrateInitialDataForWidgetVisibility(initialData)
    }
    getActions().setInitialData(initialData)
    ...
```

```ts
// migrate_initial_data.ts or initial_data.ts
export async function migrateInitialDataForWidgetVisibility(initialData: InitialData) {
  // Other widget's auth state can be fetched from local storage.
  // But, need to fetch ftx auth state to migrate.
  // Start migration when ftx auth state is ready.
  const ftxUserAuthed = new Promise(resolve => chrome.ftx.getAccountBalances((balances, authInvalid) => {
    resolve(!authInvalid)
  })
  const peristentState: NewTab.PersistentState = storage.load()
  const widgetLookupTable = {
    'braveTalk': {
      display: initialData.braveTalkSupported && initialData.preferences.showBraveTalk,
      isCrypto: false
    },
    'rewards': {
      display: initialData.preferences.showRewards,
      isCrypto: false
    },
    'binance': {
      display: initialData.binanceSupported && initialData.preferences.showBinance,
      isCrypto: true,
      userAuthed: peristentState.binanceState.userAuthed
    },
    'cryptoDotCom': {
      display: initialData.cryptoDotComSupported && initialData.preferences.showCryptoDotCom,
      isCrypto: true,
      userAuthed: false
    },
    'ftx': {
      display: initialData.ftxSupported && initialData.preferences.showFTX,
      isCrypto: true,
      userAuthed: ftxUserAuthed
    },
    'gemini': {
      display: initialData.geminiSupported && initialData.preferences.showGemini,
      isCrypto: true,
      userAuthed: peristentState.geminiState.userAuthed
    }
  }

  // Find crypto widget that is foremost and visible.
  let foremostVisibleCryptoWidget = ''
  const lastIndex = peristentState.widgetStackOrder.length - 1
  for (let i = lastIndex; i >= 0; --i) {
    const widget = widgetLookupTable[peristentState.widgetStackOrder[i]]
    if (!widget) {
      console.error('Update above lookup table')
      continue
    }

    if (!widget.display) {
      continue
    }

    if (widget.isCrypto) {
      foremostVisibleCryptoWidget = peristentState.widgetStackOrder[i]
    }
    // Found visible foremost widget in the widget stack. Go out.
    break
  }

  const widgetsShowState = {
    'binance': false,
    'cryptoDotCom': false,
    'ftx': false,
    'gemini': false
  }

  for (const key in widgetsShowState) {
    // Show foremost visible crypto widget regardless of auth state
    // and show user authed crypto widget.
    if (key === foremostVisibleCryptoWidget ||
        widgetLookupTable[key].userAuthed) {
      widgetsShowState[key] = true
    }
  }

  // These don't return promise so we can't await and then fetch new preferences,
  // instead make manual changes to initialData.preferences after.
  saveShowBinance(widgetsShowState.binance)
  saveShowCryptoDotCom(widgetsShowState.cryptoDotCom)
  saveShowFTX(widgetsShowState.ftx)
  saveShowGemini(widgetsShowState.gemini)
  saveWidgetVisibilityMigrated()

  initialData.preferences = {
    ...initialData.preferences,
    showBinance: widgetsShowState.binance,
    showCryptoDotCom: widgetsShowState.cryptoDotCom,
    showFTX: widgetsShowState.ftx,
    showGemini: widgetsShowState.gemini
  }
}
```

Reply:
Thanks @petemill ! Addressed by https://github.com/brave/brave-core/pull/11266/commits/ba78700735508ee10b44bd389133583da01ae56d